stdlib/domain.mli

```diff
@@ -90,15 +90,38 @@ module DLS : sig
 
     val new_key : ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a key
     (** [new_key f] returns a new key bound to initialiser [f] for accessing
-        domain-local variables.
-
-        If [split_from_parent] is provided, spawning a domain will derive the
-        child value (for this key) from the parent value.
-
-        Note that the [split_from_parent] call is computed in the parent
-        domain, and is always computed regardless of whether the child domain
-        will use it. If the splitting function is expensive or requires
-        client-side computation, consider using ['a Lazy.t key].
+,        domain-local variables.
+
+        If [split_from_parent] is not provided, the value for a new
+        domain will be computed on-demand by the new domain: the first
+        [get] call will call the initializer [f] and store that value.
+
+        If [split_from_parent] is provided, spawning a domain will
+        derive the child value (for this key) from the parent
+        value. This computation happens in the parent domain and it
+        always happens, regardless of whether the child domain will
+        use it.
+        If the splitting function is expensive or requires
+        child-side computation, consider using ['a Lazy.t key]:
+
+        {[
+        let init () = ...
+
+        let split_from_parent parent_value =
+          ... parent-side computation ...;
+          lazy (
+            ... child-side computation ...
+          )
+
+        let key = Domain.DLS.new_key ~split_from_parent init
+
+        let get () = Lazy.force (Domain.DLS.get key)
+        ]}
+
+        In this case a part of the computation happens on the child
+        domain; in particular, it may access [parent_value]
+        concurrently with the parent domain, which requires careful
+        programming.
```

Comment:
I am not sure what "careful programming" means here. I would propose to remove this proposition.

Reply:
I'm trying to say that it's the programmer responsibility to ensure that this concurrent access is safe -- there is no synchronization done in the runtime. Maybe removing ", which requires careful programming" is clear enough, but I was worried that "may" could be interpreted as "you have the permission to do this", implying that we make it magically safe.

Reply:
To keep with the spirits of the warnings and alerts on unsynchronized accesses, what about
```
which may require explicit synchronization to avoid data races. 
```
?

Reply:
Thanks! I went with your suggestion.